src/server/chunk_server/chunk_server_lease_service_impl.h

```diff
@@ -0,0 +1,32 @@
+#ifndef GFS_SERVER_CHUNK_SERVER_CHUNK_SERVER_LEASE_SERVICE_IMPL_H_
+#define GFS_SERVER_CHUNK_SERVER_CHUNK_SERVER_LEASE_SERVICE_IMPL_H_
+
+#include "grpcpp/grpcpp.h"
+#include "src/protos/grpc/chunk_server_lease_service.grpc.pb.h"
+
+namespace gfs {
+namespace service {
+
+// The synchronous implementation for handling ChunkServerLeaseService requests
+class ChunkServerLeaseServiceImpl final
```

Comment:
The master is the one that grants and revokes the lease, but the master sends these gRPC requests to a port that the chunk servers listen on, so the chunk servers can act on it. In the gRPC scheme, the "client" is the one who sends the requests and the "server" is the one that responds to the request. Since the master are the ones that "send lease gRPC" to the chunk servers,  this is the server side logic for "chunks server" to handle the master's requests.

Reply:
And yes, when the master chooses a primary replica, it sends a "GrantLease" gRPC to the primary chunk server, and that chunk server's ChunkServerLeaseServiceImpl will handle the request and use the lease, if applicable

Reply:
Yes, I think the gRPC server should be on the master and the client should be the chunk server. Also apart from the master being the one granting lease, the chunkserver needs to periodically renew the lease from the master or else the lease will expire. So it needs to make calls to the master to renew the lease. For example if a chunkserver goes down, it will be unable to renew the lease and the lease will expire and master can give it to another chunkserver. I may be missing something here, but lets continue this conversation on slack or maybe a call later.